File: RBE/16_traits.cpp
#include <string>
#include <iostream>

using namespace std;

template<typename Self>
struct Animal
{
    // Associated function signature; `Self` refers to the implementor type.
    static Self create(const string& name);

    // Method signatures; these will return a string.
    const string& get_name();
    const string& get_noise();

    // Traits can provide default method definitions.
    void talk() {
        cout << this->get_name() << " says " << this->get_noise() << endl;
    }
};

struct Sheep: Animal<Sheep>
{
    bool is_naked() {
        return this->naked;
    }

    void shear() {
        if (is_naked()) {
            // Implementor methods can use the implementor's trait methods.
            cout << this->get_name() << " is already naked...\n";
        } else {
            cout << this->name << " gets a haircut!\n";

            this->naked = true;
        }
    }

    // Implement the `Animal` trait for `Sheep`
    static Sheep create(const string& name) {
        return Sheep{name, false};
    }

    const string& get_name() {
        return this->name;
    }

    const string& get_noise() {
        if (this->is_naked()) {
            return Sheep::noise_naked;
        } else {
            return Sheep::noise;
        }
    }

    // Default trait methods can be overridden.
    void talk() {
        // For example, we can add some quiet contemplation.
        cout << this->name << " pauses briefly... " << this->get_noise() << endl;
    }

    bool naked;
    const string& name;

    static const string noise;
    static const string noise_naked;

private:
    Sheep(const string& name, bool naked):
        name(name), naked(naked) {}
};

const string Sheep::noise = string{"baaaaah!"};
const string Sheep::noise_naked = string{"baaaaah?"};

int main(void)
{
    auto dolly = Sheep::create("Dolly");

    dolly.talk();
    dolly.shear();
    dolly.talk();

    return 0;
}

File: RBE/16_traits.rs
struct Sheep { naked: bool, name: &'static str }

trait Animal {
    // Associated function signature; `Self` refers to the implementor type.
    fn create(name: &'static str) -> Self;

    // Method signatures; these will return a string.
    fn get_name(&self) -> &'static str;
    fn get_noise(&self) -> &'static str;

    // Traits can provide default method definitions.
    fn talk(&self) {
        println!("{} says {}", self.get_name(), self.get_noise());
    }
}

impl Sheep {
    fn is_naked(&self) -> bool {
        self.naked
    }

    fn shear(&mut self) {
        if self.is_naked() {
            // Implementor methods can use the implementor's trait methods.
            println!("{} is already naked...", self.get_name());
        } else {
            println!("{} gets a haircut!", self.name);

            self.naked = true;
        }
    }
}

// Implement the `Animal` trait for `Sheep`.
impl Animal for Sheep {
    // `Self` is the implementor type: `Sheep`.
    fn create(name: &'static str) -> Sheep {
        Sheep { name: name, naked: false }
    }

    fn get_name(&self) -> &'static str {
        self.name
    }

    fn get_noise(&self) -> &'static str {
        if self.is_naked() {
            "baaaaah?"
        } else {
            "baaaaah!"
        }
    }

    // Default trait methods can be overridden.
    fn talk(&self) {
        // For example, we can add some quiet contemplation.
        println!("{} pauses briefly... {}", self.name, self.get_noise());
    }
}

fn main() {
    // Type annotation is necessary in this case.
    let mut dolly: Sheep = Animal::create("Dolly");
    // TODO ^ Try removing the type annotations.

    dolly.talk();
    dolly.shear();
    dolly.talk();
}
